src/vehicle.cpp

```diff
@@ -5704,190 +5705,180 @@ bool vehicle::shift_if_needed() {
-                        explosion_chance = 2;
-                    }
-                    fiery_explosion = false; // WOOF!!; but no lasting flames
-                    fuel_size_factor = 1;
-                    explosion_factor = 1.4; // Higher energy density, but dampened by the explosion type
-                }
-                const int pow = 120 * (1 - exp(explosion_factor / -5000 * (parts[p].amount * fuel_size_factor)));
-                //debugmsg( "damage check dmg=%d pow=%d amount=%d", dmg, pow, parts[p].amount );
-                if(parts[p].hp <= 0) {
-                    leak_fuel( p );
-                }
-                if (one_in(explosion_chance)) {
-                    g->u.add_memorial_log(pgettext("memorial_male","The fuel tank of the %s exploded!"),
-                        pgettext("memorial_female", "The fuel tank of the %s exploded!"),
-                        name.c_str());
-                    g->explosion( tripoint( global_x() + parts[p].precalc[0].x,
-                                            global_y() + parts[p].precalc[0].y,
-                                            smz ),
-                                  pow, 0.7, fiery_explosion );
-                    parts[p].hp = 0;
-                    parts[p].amount = 0;
-                }
-            }
-        } else if (parts[p].hp <= 0 && part_flag(p, "UNMOUNT_ON_DAMAGE")) {
-            tripoint dest( global_x() + parts[p].precalc[0].x,
-                           global_y() + parts[p].precalc[0].y,
-                           smz );
-            g->m.spawn_item( dest, part_info(p).item, 1, 0, calendar::turn );
-            remove_part( p );
+
+    const fuel_explosion &data = iter->second;
+    const int pow = 120 * (1 - exp(data.explosion_factor / -5000 * (parts[p].amount * data.fuel_size_factor)));
+    //debugmsg( "damage check dmg=%d pow=%d amount=%d", dmg, pow, parts[p].amount );
+    if( parts[p].hp <= 0 ) {
+        leak_fuel( p );
+    }
+
+    int explosion_chance = type == DT_HEAT ? data.explosion_chance_hot : data.explosion_chance_cold;
+    if( one_in( explosion_chance ) ) {
+        g->u.add_memorial_log(pgettext("memorial_male","The fuel tank of the %s exploded!"),
+            pgettext("memorial_female", "The fuel tank of the %s exploded!"),
+            name.c_str());
+        g->explosion( global_part_pos3( p ), pow, 0.7, data.fiery_explosion );
+        parts[p].hp = 0;
+        parts[p].amount = 0;
+    }
+
+    return true;
+}
+
+int vehicle::damage_direct( int p, int dmg, damage_type type )
+{
+    if( parts[p].hp <= 0 ) {
+        return break_off( p, dmg );
+    }
+
+    int tsh = std::min( 20, part_info(p).durability / 10 );
+    if( dmg < tsh && type != DT_TRUE ) {
+        if( type == DT_HEAT && part_flag( p, "FUEL_TANK" ) ) {
+            explode_fuel( p, type );
         }
+
+        return dmg;
     }
-    if (dres < 0)
-        dres = 0;
-    return dres;
+
+    dmg -= part_info( p ).damage_reduction[ type ];
```

Comment:
This could lead to negative values for `dmg`, leading to an actual increase of hp of the part.